File: LeetCode/Cpp/[1] Two Sum.cpp
// 1. Reflection

// - Hash table made it so simple.
// - Try to use hash table once it's possible.

// 2. Problem

// Given an array of integers, find two numbers such that they add up to a specific target number.
// The function twoSum should return indices of the two numbers such that they add
// up to the target, where index1 must be less than index2. Please note that your
// returned answers (both index1 and index2) are not zero-based.
// You may assume that each input would have exactly one solution.

// 3. Initial solution( complexity: time=>O(n^2) space=>O(1))

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        for(int i=0; i< nums.size(); i++){
            for(int j=0; j < nums.size(); j++){
                if(i == j) continue;
                else if(nums[i] + nums[j] == target){
                    return {i, j};
                }
            }
        }
    }
};

// 4. Improved solution( complexity: time=>O(n) space=>O(n))

class Solution {
public:
    vector<int> twoSum(vector<int> &numbers, int target){
        unordered_map<int, int> hash;
        for(int i=0; i<numbers.size(); i++){
            int numNeed = target - numbers[i];
            if(hash.find(numNeed) != hash.end()){
                return { hash[numNeed], i};
            }
            hash[numbers[i]] = i;
        }
    }
};

File: LeetCode/Javascript/[1] Two Sum.js
// 1. Reflection

// - Solved.
// - Use hasOwnProperty when I need to check whether prop in Object. It's faster,
//   because it doesn't need to check protypical chain.

// 2. Problem

// Given an array of integers, return indices of the two numbers such that
// they add up to a specific target.
//
// You may assume that each input would have exactly one solution,
// and you may not use the same element twice.

// 3. Initial solution( complexity: time=>O() space=>O())

// complexity: time=>O(N^2) space=>O(1)
var twoSum = function(nums, target) {
  for (let i = 0; i < nums.length; i++) {
    for (let j = 0; j < nums.length; j++) {
      if (i !== j && nums[i] + nums[j] === target) return [i, j];
    }
  }
};

// complexity: time=>O(N^2) space=>O(1))
var twoSum = function(nums, target) {
  for (let i = 0; i < nums.length; i++) {
    let val = target - nums[i];
    let idx = nums.lastIndexOf(val);
    if (idx > 0 && idx !== i) return [i, idx];
  }
};

// 4. Improved solution( complexity: time=>O(N) space=>O(N) )

// 92ms
var twoSum = function(nums, target) {
  let duplicate = {};
  for (let i = 0; i < nums.length; i++) {
    if (duplicate.hasOwnProperty(nums[i])) return [duplicate[nums[i]], i];
    duplicate[target - nums[i]] = i;
  }
};

//112ms
var twoSum = function(nums, target) {
  let duplicate = {};
  for (let i = 0; i < nums.length; i++) {
    if (duplicate[nums[i]] !== undefined) return [duplicate[nums[i]], i];
    duplicate[target - nums[i]] = i;
  }
};
